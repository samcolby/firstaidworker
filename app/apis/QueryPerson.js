import React from "react";
import PropTypes from "prop-types";

import { Query } from "react-apollo";

// Generate the GQL queries used in this component
import { QUERY_PERSON } from "../GraphQLQueries";

/**
 * Apollo Query to return the data for a person based on their id
 * @param {Object} props - Standard react prop
 * @param {function} children - Standard react prop, doesn't need to be included
 * @param {string} id - the id of the person to retrieve
 */
function QueryPerson({ children, id }) {
  return (
    <Query query={QUERY_PERSON} variables={{ id }}>
      {props => children(props)}
    </Query>
  );
}

QueryPerson.propTypes = {
  children: PropTypes.func,
  id: PropTypes.string.isRequired
};

export default QueryPerson;
