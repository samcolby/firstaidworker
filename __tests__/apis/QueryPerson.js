import React from "react";
import { Text } from "react-native";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import { MockedProvider } from "react-apollo/test-utils";

import wait from "waait";

import QueryPerson, { QUERY_PERSON_DATA } from "../../app/apis/QueryPerson";

import {
  mocksForQueryPerson,
  mocksForQueryPersonError
} from "../mocksdata/MocksQueryPerson";

const ID = "54f58dfd-3f17-427c-b1d1-74c26a77f700";

describe("<QueryPerson /> component", () => {
  it("Renders correctly", () => {
    const component = shallow(<QueryPerson id={ID} />);
    expect(component.length).toBe(1);
  });

  it("Renders loading correctly", () => {
    const component = renderer.create(
      <MockedProvider mocks={mocksForQueryPerson} addTypename={false}>
        <QueryPerson id={ID}>
          {({ data, error, fetchMore, loading, networkStatus, refetch }) => {
            if (error) {
              return <Text>error</Text>;
            } else if (loading) {
              return <Text>Loading</Text>;
            } else {
              return <Text>{JSON.stringify(data[QUERY_PERSON_DATA])}</Text>;
            }
          }}
        </QueryPerson>
      </MockedProvider>
    );

    const textComponent = component.root.findByType("Text");
    expect(textComponent.children[0]).toEqual("Loading");
    expect(component).toMatchSnapshot();
  });

  it("Renders data correctly", async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocksForQueryPerson} addTypename={false}>
        <QueryPerson id={ID}>
          {({ data, error, fetchMore, loading, networkStatus, refetch }) => {
            if (error) {
              return <Text>Error: {JSON.stringify(error)}</Text>;
            } else if (loading) {
              return <Text>loading</Text>;
            } else {
              return <Text>{JSON.stringify(data[QUERY_PERSON_DATA])}</Text>;
            }
          }}
        </QueryPerson>
      </MockedProvider>
    );

    await wait(0);

    const textComponent = component.root.findByType("Text");
    expect(JSON.parse(textComponent.children)[0].name).toEqual("Key Whitley");

    expect(component).toMatchSnapshot();
  });

  it("Throws errors", async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocksForQueryPersonError} addTypename={false}>
        <QueryPerson id={ID}>
          {({ data, error, fetchMore, loading, networkStatus, refetch }) => {
            if (error) {
              return <Text>ERROR!</Text>;
            } else if (loading) {
              return <Text>loading</Text>;
            } else {
              return <Text>{JSON.stringify(data[QUERY_PERSON_DATA])}</Text>;
            }
          }}
        </QueryPerson>
      </MockedProvider>
    );

    await wait(0); // wait for response

    const textComponent = component.root.findByType("Text");
    expect(textComponent.children[0]).toContain("ERROR!");
  });
});
