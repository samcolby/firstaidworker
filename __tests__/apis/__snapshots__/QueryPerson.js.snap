// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QueryPerson /> component Renders data correctly 1`] = `
<Text>
  [{"id":"54f58dfd-3f17-427c-b1d1-74c26a77f700","name":"Key Whitley","job_title":"Michael Jackson Impersonator","is_active":true,"avatar_uri":"https://randomuser.me/api/portraits/women/45.jpg","company":{"name":"DIGINETIC"},"department":"Information Technology","email":"keywhitley@diginetic.com","phone":"+1 (968) 469-2764","location":{"coordinates":[-78.5011,-35.033]}}]
</Text>
`;

exports[`<QueryPerson /> component Renders loading correctly 1`] = `
<Text>
  Loading
</Text>
`;
